app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/OfflineRegionDownloadActivity.kt

```diff
@@ -0,0 +1,184 @@
+
+import android.graphics.Color
+import android.graphics.PointF
+import android.os.Bundle
+import android.support.v7.app.AppCompatActivity
+import android.view.View
+import android.widget.ArrayAdapter
+import android.widget.Toast
+import com.mapbox.api.routetiles.v1.versions.models.RouteTileVersionsResponse
+import com.mapbox.geojson.BoundingBox
+import com.mapbox.mapboxsdk.Mapbox
+import com.mapbox.mapboxsdk.maps.MapboxMap
+import com.mapbox.mapboxsdk.style.layers.FillLayer
+import com.mapbox.mapboxsdk.style.layers.PropertyFactory.fillColor
+import com.mapbox.mapboxsdk.style.sources.GeoJsonSource
+import com.mapbox.services.android.navigation.testapp.R
+import com.mapbox.services.android.navigation.v5.navigation.OfflineTileVersions
+import com.mapbox.services.android.navigation.v5.navigation.OfflineTiles
+import com.mapbox.services.android.navigation.v5.navigation.RoutingTileDownloadManager
+import kotlinx.android.synthetic.main.activity_offline_region_download.*
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+
+class OfflineRegionDownloadActivity : AppCompatActivity(), RoutingTileDownloadManager.RoutingTileDownloadListener {
+
+    lateinit var mapboxMap: MapboxMap
+    private val disabledGrey by lazy { resources.getColor(R.color.md_grey_700) }
+    private val enabledBlue by lazy { resources.getColor(R.color.mapbox_blue) }
+    private var downloadButtonEnabled: Boolean = false
+
+    private val boundingBox: BoundingBox
+        get() {
+            val top = selectionBox.top - mapView.top
+            val left = selectionBox.left - mapView.left
+            val right = left + selectionBox.width
+            val bottom = top + selectionBox.height
+
+            val southWest = mapboxMap.projection.fromScreenLocation(
+                    PointF(left.toFloat(), bottom.toFloat()))
+            val northEast = mapboxMap.projection.fromScreenLocation(
+                    PointF(right.toFloat(), top.toFloat()))
+
+            return BoundingBox.fromLngLats(
+                    southWest.longitude, southWest.latitude,
+                    northEast.longitude, northEast.latitude)
+        }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_offline_region_download)
+
+        downloadButton.setOnClickListener { onDownloadClick() }
+        setupSpinner()
+        setupMapView(savedInstanceState)
+    }
+
+    fun setupSpinner() {
+        OfflineTileVersions(Mapbox.getAccessToken())
+                .getRouteTileVersions(object : Callback<RouteTileVersionsResponse> {
+                    override fun onResponse(call: Call<RouteTileVersionsResponse>, response:
+                    Response<RouteTileVersionsResponse>) {
+                        setupSpinner(response.body()!!.availableVersions())
+                    }
+
+                    override fun onFailure(call: Call<RouteTileVersionsResponse>, throwable: Throwable) {
```

Comment:
Should we do something here? Seems like an unhandled failure would result in a dead end for the user, even though this is just an example... could we show a simple error and re-enable downloads?

Reply:
@akitchen I added a recovery button for when versions are not loaded which allows the user to try to re-fetch the versions.